rfcs/20200826-keras-nlp-scoping-design.md

```diff
@@ -0,0 +1,108 @@
+# Keras NLP
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Mark Omernick (momernick@google.com), Francois Chollet (fchollet@google.com) |
+| **Updated**   | 2020-08-26                                           |
+
+
+## Objective
+
+We aim at describing the scope of [keras-nlp](https://github.com/keras-team/keras-nlp), especially:
+
+- What use cases `keras-nlp` should cover
+- Boundaries between `keras-nlp` and [tensorflow addons](https://github.com/tensorflow/addons)
+- Boundaries between `keras-nlp` and [tensorflow model garden](https://github.com/tensorflow/models)
+- Boundaries between `keras-nlp` and [tf.keras](https://www.tensorflow.org/api_docs/python/tf/keras).
+- Boundaries between `keras-nlp` and [tf.text](https://www.tensorflow.org/tutorials/tensorflow_text/intro).
+
+## Motivation
+
+Natural Language Processing (NLP) is a major application area for our users.
+In recent years, Transformer-based models have become the foundation of many NLP workflows.
+These workflows tend to reuse similar components, for which in some cases third-party packages
+have been developed by the open-source community.
+
+These third-party solutions are not always kept up to date or up to the same quality standards as core Keras.
+They also raise the issue of API standardization.
+
+To fix this, we want machine learning engineers to have access to a standard Keras-native,
+optimized, and well-tested set of components to build their Transformer-based (and beyond) NLP workflows.
+
+This provides key user benefits:
+
+- The package would be first-party and thus always up to date with modern best practices.
+- High code quality and testing standards and strict quality control: same level of trust as core Keras
+- A shared API standard across the community
+- Ability for the open-source community to build more advanced solutions *on top* of this package instead of reinventing it
+- Ability for research scientists to benefit from subclassing and customizing base components to quickly test new research ideas
+
+## Design Proposal
+
+`keras-nlp` will include most standard Transformer-based modules, specifically:
+
+- Keras layer components such as Transformer encoder and decoder blocks.
+- Keras task components such as masked language, span labeler and named entity recognition.
+- Tensorflow operations such as beam search.
```

Comment:
Does this scope include custom op kernels for things such as beam search? [TFA seq2seq uses one](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/custom_ops/seq2seq/cc/kernels/beam_search_ops.cc) but that would seem to differ from the python only Keras repositories that I thought were being created.

Reply:
Same as Keras core that I would like this to be pure python library, or at least for the current scoping. TF has mechanism to support things that cannot be interpreted as pure numpy ops.
Specifically, for beam search, I wonder if this would be something tf.text can take -- worth discussing between us.